File: lincoln/ruby/lib/types/table/ref_table.rb
require "lib/types/table/hash_index"
require "lib/types/exception/bad_key_exception"
require "lib/types/table/table"

class RefTable < Table
  def initialize(context, name, key, types)
    super(name, TableType::TABLE, key, types)
    @key = key
    @tuples = TupleSet.new(name, nil)
    @primary = HashIndex.new(context, self, key, Index::Type::PRIMARY)
    @secondary = Hash.new
  end

  def tuples
    return (@tuples.nil? ? TupleSet.new(name) : @tuples)
  end
  
  attr_reader :primary, :secondary

  def cardinality
    @tuples.size
  end

  protected  
    def delete_tup(t)
      if @tuples.include?(t) then
        lookup = primary.lookup(t)
      end

      return nil if lookup.nil?
      lookup.each do |l|
        if l == t then
          if l.count > 1.0 then
            l.count = l.count - 1.0
            return nil
          end
          return tuples.delete(t)
        end
      end
      return nil
    end

    def insert_tup(t)
      if @tuples.include?(t) then
        primary.lookup(t).each do |l|
          if l == t then
            l.count = l.count + 1.0
            return false
          end
        end
      end
      @tuples << t.clone
      return true
    end

end

File: lincoln/java/src/jol/types/table/RefTable.java
package jol.types.table;

import java.util.Hashtable;
import java.util.Iterator;

import jol.core.Runtime;
import jol.types.basic.Tuple;
import jol.types.basic.TupleSet;
import jol.types.basic.TypeList;
import jol.types.exception.BadKeyException;
import jol.types.exception.UpdateException;

/**
 * A RefCount Table (RefTable).
 * Tables of this type maintain a reference count on
 * each stored tuple. Insertions into this table create
 * a new tuple with a refcount of 1 or increment the refcount
 * of a preexisting tuple. Deletions into this table decrement
 * the refcount of the deleted tuple and remove that tuple if
 * the refcount goes to 0.
 *
 */
public class RefTable extends Table {
	/** The set of tuples stored by this table instance. */
	protected TupleSet tuples;
	
	/** A reference to the primary index. */
	protected Index primary;
	
	/** A map to all secondary indices. */
	protected Hashtable<Key, Index> secondary;
	
	/**
	 * Create a new RefCounted Table.
	 * @param context The runtime context.
	 * @param name The name of the table.
	 * @param key The primary key.
	 * @param types The list of attribute types.
	 */
	public RefTable(Runtime context, TableName name, Key key, TypeList types) {
		super(name, Type.TABLE, key, types);
		this.tuples = new TupleSet(name);
		this.primary = new HashIndex(context, this, key, Index.Type.PRIMARY);
		this.secondary = new Hashtable<Key, Index>();
	}
	
	@Override
	public Iterator<Tuple> tuples() {
		return (this.tuples == null ? new TupleSet(name()) : this.tuples.clone()).iterator();
	}
	
	@Override
	protected boolean delete(Tuple t) throws UpdateException {
		if (this.tuples.contains(t)) {
			TupleSet lookup;
			try {
				lookup = primary().lookupByKey(primary().key().project(t));
			} catch (BadKeyException e) {
				throw new UpdateException(e.toString());
			}
		
			for (Tuple l : lookup) {
				if (l.equals(t)) {
					if (l.refCount() > 1L) {
						l.refCount(l.refCount() - 1L);
						return false; // The tuple has not been deleted yet.
					}
					return this.tuples.remove(t);
				}
			}
		}
		return false; // The tuple does not exist.
	}


	@Override
	protected boolean insert(Tuple t) throws UpdateException {
		if (this.tuples.contains(t)) {
			try {
				for (Tuple lookup : primary().lookupByKey(primary().key().project(t))) {
					if (lookup.equals(t)) {
						lookup.refCount(lookup.refCount().longValue() + 1);
						return false; 
					}
				}
			} catch (BadKeyException e) {
				throw new UpdateException(e.toString());
			}
		}
		this.tuples.add(t.clone());
		return true;
	}

	@Override
	public Index primary() {
		return this.primary;
	}

	@Override
	public Hashtable<Key, Index> secondary() {
		return this.secondary;
	}

	@Override
	public Integer cardinality() {
		return this.tuples.size();
	}

}
